src/app/warehouse/warehouse.cpp

```diff
@@ -0,0 +1,74 @@
+#include "warehouse.hpp"
+#include "product.hpp"
+
+void Warehouse::increaseAmount(const std::shared_ptr<Product> &ptrToProduct, const uint amount)
+{
+    for (auto &product_ : products_)
+    {
+        if (product_ == ptrToProduct)
+        {
+            product_->increaseAmount(amount);
+        }
+    }
+}
+
+void Warehouse::decreaseAmount(const std::shared_ptr<Product> &ptrToProduct, uint amount)
+{
+    for (auto &product_ : products_)
+    {
+        if (product_ == ptrToProduct)
+        {
+            if (amount > product_->getAmount())
+            {
+                product_->setAmount(0);
+                return;
+            }
+            else
+                product_->decreaseAmount(amount);
+        }
+    }
+}
+
+void Warehouse::addProducts(const std::vector<std::shared_ptr<Product>> &products)
+{
+    for (const auto &product : products)
+    {
+        if (std::ranges::none_of(products_, [&](const auto &ptrToProduct) { return ptrToProduct == product; }))
```

Comment:
According to std::set::insert() documentation, insert "Inserts element(s) into the container, if the container doesn't already contain an element with an equivalent key."

So, after good change to set simple insert instruction below should be enough.